File: utils/bench.nim
import os, osproc, strutils, strformat, sequtils

let num = parseInt(paramStr(1))

let rust = &"./Rust/bin/Euler{num}"
let haskell = &"./Haskell/bin/{num}"
var nim = &"./Nim/bin/p00{num}"

case num
of 1..9: nim = &"./Nim/bin/p00{num}"
of 10..100: nim = &"./Nim/bin/p0{num}"
else: nim = &"./Nim/bin/p{num}"

proc main =
  let output = case num
  of 1..9: execProcess(&"hyperfine -w 3 --export-markdown Bench/euler00{num}.md " & rust & " " & nim & " " & haskell)
  of 10..99: execProcess(&"hyperfine -w 3 --export-markdown Bench/euler0{num}.md " & rust & " " & nim & " " & haskell)
  else: execProcess(&"hyperfine -w 3 --export-markdown Bench/euler{num}.md " & rust & " " & nim & " " & haskell)

  echo output

if isMainModule:
  main()

File: bench.jl
using DataFrames

strnum, strtimes = ARGS
num = parse(Int64, strnum)
times = parse(Int64, strtimes)

rust = `./Rust/bin/Euler$(num)`
nim = `./Nim/bin/p00$(num)`
haskell = `./Haskell/bin/$(num)`

rust_time = 0
nim_time = 0
haskell_time = 0

for i = 1:times
  rust_time += @elapsed run(rust);
  nim_time += @elapsed run(nim);
  haskell_time += @elapsed run(haskell);
end

rust_time /= times
nim_time /= times
haskell_time /= times

println("Rust: ", rust_time)
println("Nim: ", nim_time)
println("Haskell: ", haskell_time)

result = DataFrame(Rust = rust_time, Nim = nim_time, Haskell = haskell_time)

if num < 10
  writetable("Bench/euler00$(num).csv", result)
elseif num < 100
  writetable("Bench/euler0$(num).csv", result)
else
  writetable("Bench/euler$(num).csv", result)
end
